File: src/main/java/j/chain_of_responsibility/inheritance/supports/LimitSupport.java
package j.chain_of_responsibility.inheritance.supports;

import j.chain_of_responsibility.Trouble;
import j.chain_of_responsibility.inheritance.Support;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class LimitSupport extends Support {
  private final int limit; // この番号未満なら解決できる

  public LimitSupport(@NotNull String name, int limit, @Nullable Support next) {
    super(name, next);
    this.limit = limit;
  }

  public LimitSupport(@NotNull String name, int limit) {
    this(name, limit, null);
  }

  @Override
  protected boolean resolve(Trouble trouble) {
    return trouble.getNumber() < limit;
  }
}

File: src/main/scala/s/chain_of_responsibility/inheritance/supports/LimitSupport.scala
package s.chain_of_responsibility.inheritance.supports

import s.chain_of_responsibility.Trouble
import s.chain_of_responsibility.inheritance.Support

final class LimitSupport(
    name: String,
    private val limit: Int,
    next: Option[Support]
) extends Support(name, next) {
  def this(name: String, limit: Int) = {
    this(name, limit, None)
  }

  override protected def resolve(trouble: Trouble): Boolean =
    trouble.number < limit
}
